index.ts

```diff
@@ -1,60 +1,83 @@
 const setImmediate = global.setImmediate || process.nextTick.bind(process);
 const noop = () => {};
 
+enum AnalyticsState {
+  Idle = 'idle',
+  Running = 'running',
+}
+
+type AnalyticsMessage = any;
+
 class Analytics {
+  private readonly enable: boolean;
+  private state = AnalyticsState.Idle;
+  private queue = [] as {
+    message: AnalyticsMessage;
+    callback: (error?: Error, data?: { batch: AnalyticsMessage[]; sentAt: Date }) => void;
+  }[];
+  private writeKey: string;
+  private host: string;
+  private timeout: number | null;
+  private flushAt: number;
+  private flushInterval: number;
+  private maxInternalQueueSize: number;
+  private flushed: boolean = false;
+  private flushTimer: NodeJS.Timer | null = null;
+  private timer: NodeJS.Timer | null = null;
+
+  private logger: bunyan;
+
   /**
-   * Initialize a new `Analytics` with your Segment project's `writeKey` and an
-   * optional dictionary of `options`.
-   *
-   * @param {String} writeKey
-   * @param {String} dataPlaneURL
-   * @param {Object=} options (optional)
-   * @param {Number=20} options.flushAt (default: 20)
-   * @param {Number=20000} options.flushInterval (default: 20000)
-   * @param {Boolean=true} options.enable (default: true)
-   * @param {Number=20000} options.maxInternalQueueSize
+   * Initialize a new `Analytics` instance with your RudderStack project's `writeKey` and an
+   * optional dictionary of options.
    */
+  constructor(
+    writeKey: string,
+    dataPlaneURL: string,
+    {
+      enable = true,
+      timeout = 0,
+      flushAt = 20,
+      flushInterval = 20000,
+      maxInternalQueueSize = 20000,
+      logLevel = bunyan.FATAL,
+    }: {
+      enable?: boolean;
+      timeout?: number;
+      flushAt?: number;
+      flushInterval?: number;
+      maxInternalQueueSize?: number;
+      logLevel?: bunyan.LogLevel;
+    } = {}
+  ) {
+    assert(writeKey, `You must pass your project's write key.`);
+    assert(dataPlaneURL, `You must pass your data plane URL.`);
 
-  constructor(writeKey, dataPlaneURL, options) {
-    options = options || {};
-
-    assert(writeKey, "You must pass your project's write key.");
-    assert(dataPlaneURL, 'You must pass your data plane url.');
-
-    this.queue = [];
-    this.state = 'idle';
     this.writeKey = writeKey;
-    this.host = removeSlash(dataPlaneURL);
-    this.timeout = options.timeout || false;
-    this.flushAt = Math.max(options.flushAt, 1) || 20;
-    this.flushInterval = options.flushInterval || 20000;
-    this.maxInternalQueueSize = options.maxInternalQueueSize || 20000;
-    this.logLevel = options.logLevel || 'fatal';
-    this.flushed = false;
-    Object.defineProperty(this, 'enable', {
-      configurable: false,
-      writable: false,
-      enumerable: true,
-      value: typeof options.enable === 'boolean' ? options.enable : true,
-    });
+    this.host = removeTrailingSlash(dataPlaneURL);
+    this.enable = enable;
```

Comment:
The property is declared as readonly to simulate the behavior before when using TypeScript (though with this PR, technically it can be changed at runtime).